src/python/gem5/utils/multisim/multisim.py

```diff
@@ -0,0 +1,299 @@
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""This module contains the gem5 MultiSim framework. The gem5 MultiSim
+work functions by allowing a user to specify multiple simulations to run from
+a single gem5 config script. The MuliSim framework will then run these
+simulations in parallel using the Python multiprocessing library.
+
+The framework works by having the user add the simulators in a configuration
+via the `add_simulator`function, each with a unique, user specified, id. This
+adds the different simulations to run in parallel to a global list. The
+MultiSim framework then uses the Python multiprocessing library to run the
+simulations in parallel by loading the config script as a module in each child
+process and then selecting the simulation to run via the iid in the global set
+of scheduled simulators jobs to run.
+The only difference between the child processes is the id of the simulator.
+
+Important notes
+---------------
+
+1. You cannot load/instnatiate the simulators in the main process. You cannot
+even load the config script (i.e., `import config_script`). This means the
+config script is passed as a string referencing the config script as a modeule.
+This script is then passed to the child processes to load.
+
+2. The child processes run `simulator.run()`. Ergo, users can't specify the
+parameters passed to the `run` function (e.g., `run(max_ticks=1000)`). However,
+in such cases the `Simulator` module should be refactored to have these
+parameters set prior to calling `run`.
+
+3. The config script cannot accept parameters. It must be parameterless.
+"""
+
+import importlib
+import multiprocessing
+from pathlib import Path
+from typing import (
+    Optional,
+    Set,
+)
+
+# A global variable which __main__.py flips to `True` when multisim is run as
+# an executable module.
+module_run = False
+
+
+class SimulatorJob:
+    def __init__(self, simulator: "Simulator", id: str):
+        self._simulator = simulator
+        self._id = id
+
+    def _redirect_outputs(self):
+        """This function redirects the child process output to a subdirectory
+        in the directory specified by the `outdir` option.
+
+        This subdirectory is named after the id assigned to the simulator.
+
+        Note: This function is not yet complete. There are two missing
+        features:
+
+        1. The stdout and stderr of the child process are not redirected to the
+           subdirectory. The appear in the terminal.
```

Comment:
This feature worked with the previous implementation of multiprocessing. The *right* thing to do is to respect the `-re` parameters of the gem5 binary that is used to spawn the multiprocessing script.

Reply:
I'm honestly not sure exactly how your code worked, but the design right now does have the parent processes parameters apply to the child process. I.e, `-re` will redirect the terminal output to the m5out directory. Is this how you intended this to work?